### ISSUE TITLE
git absorb

### ISSUE THREAD

**Issue by bbigras** [magit/magit#3723]
People in the comments on https://lobste.rs/s/wj7arq/git_absorb seemed to think Magit didn't support a useful trick. Maybe it could be nice to have it.

**Comment by kyleam** [comment]
> Nice, thanks

You're welcome, but I hope you have better luck with git-autofixup than I did.  When I tried to use `magit-commit-absorb`, I quickly ran into torbiak/git-autofixup#6 :(

**Comment by marsam** [comment]
FWIW I use the following for https://github.com/tummychow/git-absorb, I haven't run into any issue so far
```elisp
;;;###autoload (autoload 'magit-commit-absorb "magit-commit" nil t)
(define-transient-command magit-commit-absorb (phase commit args)
  "Spread staged changes across recent commits.
With a prefix argument use a transient command to select infix
arguments.  This command requires git-absorb, which is available
from https://github.com/tummychow/git-absorb."
  ["Arguments"
   ("-f" "Skip safety checks"   ("-f" "--force"))
   ("-v" "Display more output"  ("-v" "--verbose"))]
  ["Actions"
   ("x"  "Absorb" magit-commit-absorb)]
  (interactive (if current-prefix-arg
                   (list 'transient nil nil)
                 (list 'select
                       (magit-get-upstream-branch)
                       (transient-args 'magit-commit-absorb))))
  (if (eq phase 'transient)
      (transient-setup 'magit-commit-absorb)
    (unless (executable-find "git-absorb")
      (user-error "This command requires the git-absorb script, which %s"
                  "is available from https://github.com/tummychow/git-absorb"))
    (unless (magit-anything-staged-p)
      (user-error "There are no staged changes that could be absorbed"))
    (when commit
      (setq commit (magit-rebase-interactive-assert commit t)))
    (if (and commit (eq phase 'run))
        (progn (magit-run-git-async "absorb" args "-b" commit) t)
      (magit-log-select
        (lambda (commit)
          (with-no-warnings ; about non-interactive use
            (magit-commit-absorb 'run commit args)))
        nil nil nil nil commit))))
```

**Comment by tarsius** [comment]
@marsam What makes you prefer `git-absorb` over `git-autofixup`?

**Comment by seanfarley** [comment]
I found this thread while trying to replace my mercurial workflow, apologies for the late reply. To answer @tarsius' question for myself: I prefer `git-absorb` over `git-autofixup` because 1) it is installable through my package manager 2) (most important to me) is that it behaves closest to `hg absorb` and doesn't require the user to select a base commit (which is a deal-breaker for me).

**Comment by tarsius** [comment]
I was about to add support for `git-absorb` out-of-the-box as well, but quickly ran into https://github.com/tummychow/git-absorb/issues/29. @seanfarley @marsam if you could help with that, I would appreciate it.

**Comment by marsam** [comment]
I think it's an issue with `TERM=dumb`, it should work if you unset `$TERM`:
```
$ TERM= git absorb
May 22 12:49:57.277 INFO committed, header: -12,1 +12,1, commit: 4c9059c48f4c892898c419d45861aa7d05b951f7
```

**Comment by tarsius** [comment]
I've added the new variant. To use it add this to your init file:

```elisp
(transient-replace-suffix 'magit-commit 'magit-commit-autofixup
  '("x" "Absorb changes" magit-commit-absorb))
```

**Comment by seanfarley** [comment]
I'm almost to my same workflow as I had with Mercurial :-) But something is very strange: why is the user prompted for choosing a base commit? Mercurial's algorithm find the base of the branch (usually the first branching point in the history).

**Comment by tarsius** [comment]
This command does something similar but there are two notable differences:

1. It uses the tip of the upstream branch as the base commit.
2. The user is given a chance to select another commit. If they choose to use the automatically detected base commit, then they can just confirm it using `C-c C-c`.

I am guessing that you tried out this command while you were on `master` and that happens to be fully merged into its upstream `origin/master`. In the base selecting dialog the cursor is placed on the first commit, which happens to be the tip of `master` and `origin/master`, but that's not so because no attempt was made to move the cursor to the appropriate base commit but because this happens to be the appropriate base commit.

**Comment by seanfarley** [comment]
Ah, not quite my situation: I was on a new branch that had not been pushed at all (so haven't set up tracking or anything remote). Having `c x` do something different than what `git absorb` does on the command-line is weird to me, but I guess the UI is up to you. I'll work around it for myself.

**Comment by tarsius** [comment]
> I was on a new branch that had not been pushed at all (so haven't set up tracking or anything remote).

Yeah, a lot of functionality is optimized around the idea that that should not happen. If you use Magit's "create a new branch", then you should almost always get one that has the *upstream* configured. The "remote where you push to" may also already be configured, but it is not the same as the upstream. See https://magit.vc/manual/magit/The-Two-Remotes.html#The-Two-Remotes and close by pages.

**Comment by seanfarley** [comment]
Oh, hmm, yeah, I mostly set up my branches with `b s` spinning off a new branch. That explains why my remote stuff wasn't set up.

Though, I still think the `-b` (base) argument should be optional (i.e. the user selects that flag before pressing `x` if they want to change the base, otherwise no commit selection screen is shown) it's not a hill I'm willing to die on :-)

**Comment by tarsius** [comment]
> I mostly set up my branches with b s spinning off a new branch. That explains why my remote stuff wasn't set up.

No it doesn't. I use that a lot too and for me that sets up the remote stuff.

> Though, I still think the `-b` (base) argument should be optional

I have tried omitting that, but so far that never worked for me. It always fails like so:

```
$ git absorb -v
May 28 22:38:05.767 WARN Please use --base to specify a base commit., line: 94, module: git_absorb::stack
May 28 22:38:05.767 CRIT No commits available to fix up, exiting, line: 27, module: git_absorb
```

And that happens even if the commit that should be selected is in the range `<upstream>..HEAD`. So yeah, that would be nice but it doesn't seem to work.

**Comment by seanfarley** [comment]
> No it doesn't. I use that a lot too and for me that sets up the remote stuff.

Huh. Then it's probably something in my setup. No worries, on this one, though.

> And that happens even if the commit that should be selected is in the range <upstream>..HEAD. So yeah, that would be nice but it doesn't seem to work.

Now that's strange but I was able to reproduce it. The `-b` argument can be omitted if there is only one ref where HEAD is. The minute I have another ref (well, let's be specific and say branch), it will error out with the same message you posted. Alas, I'm not sure there's a clean solution here without changing how `git absorb` works. Thanks for your investigation and explanation, @tarsius :-)

**Comment by tummychow** [comment]
yes, the behavior of git-absorb is to guess a base commit from "commits on your current local branch, but not any other local branch". so if you have multiple branches at the same commit, and one of them is checked out, git-absorb will ignore all of the commits, because they're shared with the other branch.

i think for a full-featured git porcelain like magit, it makes sense to completely take over and always pass the `-b` flag, rather than trusting git-absorb's best guess. magit has more information and can afford to be more opinionated about the right choice of base. i'm open to suggestions for better ways that git-absorb could infer the base (although let's talk in an issue on the git-absorb repo, rather than cluttering this one), but people use git branches in so many different ways, and there might not be a one-size-fits-all solution that's better than what i currently have. mercurial had an explicit concept of "public" and "private" branches, so hg absorb could just select all the "private" commits, but that's not a distinction we have in git

**Comment by tarsius** [comment]
IMO that logic makes sense too. And yes there others do too. :smiley_cat: 

@seanfarley I'll stick with what we have for now, i.e. *try* to guess some base commit using *some* heuristic but *always* giving the user the option to pick something else. Maybe we can tweak that in the future but this has worked quite well for me, so I see no immediate need for improvements.

**Comment by seanfarley** [comment]
Yep, that's fine. Though, I'd love to discuss more of how Mercurial does this (and correct some minor nits from @tummychow ;-)), we can do that elsewhere. And yeah, the behavior seems the most sensible for now. I'll try and mull on it some in the future :-)

**Comment by tarsius** [comment]
You can also do it here. I would like to listen in.

**Comment by seanfarley** [comment]
Ah, ok, sure, I can do that. I've been a bit distracted / preoccupied lately so I'll need a bit more time to respond.
